sql/2020/07_SEO/pages_canonicals_by_device.sql

```diff
@@ -0,0 +1,128 @@
+#standardSQL
+# page canonicals metrics grouped by device
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# returns all the data we need from _canonicals
+CREATE TEMPORARY FUNCTION get_canonicals_info(canonicals_string STRING)
+RETURNS STRUCT<
+
+  ##Antoine
+  canonicals ARRAY<STRING>, 
+  has_self_canonical BOOL, 
+  is_canonicalized BOOL,
+  http_canonicals ARRAY<STRING>, 
+  has_canonical_mismatch BOOL, 
+  canonical_raw ARRAY<STRING>, 
+  canonical_rendered ARRAY<STRING>
+
+
+
+  # add more properties here...
+
+> LANGUAGE js AS '''
+var result = {};
+try {
+    //var canonicals = JSON.parse(canonicals_string); // LIVE
+
+    // TEST
+    // Fill with test data 
+
+    var canonicals = {}; 
+
+    if (Array.isArray(canonicals) || typeof canonicals != 'object') return result;
+
+
+    if (canonicals.canonicals) {
+      result.canonicals = canonicals.canonicals;
+    }
+
+    if (canonicals.self_canonical) {
+      result.has_self_canonical = canonicals.self_canonical;
+    }
+
+    if (canonicals.other_canonical) {
+      result.is_canonicalized = canonicals.other_canonical;
+    }
+
+    if (canonicals.http_header_link_canonical) {
+      result.http_canonicals = canonicals.http_header_link_canonical;
+    }
+
+    if (canonicals.canonical_mismatch) {
```

Comment:
I spelt it wrong. canonical_mis**s**match 🤭

Reply:
Updated :)